### PR TITLE
[MacOS]Input Validation in choiceset and toggle

### PR DESCRIPTION
**Pull request by mukuagar**
## Related Issue
Please use one of the well-known [github fixes keywords](https://help.github.com/en/articles/closing-issues-using-keywords) to reference
the issues fixed with this PR (eg Fixes #<github issue number>). If an issue doesn't yet exist please create one if reasonable to aid 
in issue tracking.

**NOTE**: For multiple issues resolved by this PR use the corresponding keywords **every time** in a comma-delimited list per the reference
page above.

## Description
Create a [descriptive title and provide all of the relevant details in the description](https://chris.beams.io/posts/git-commit/). 
This information lives with the code and will help future readers (***including yourself***) and will serve as documentation.

At the very least please describe the issue you are addressing, and what your fix entails. 

Add support for input validation in choiceset (Expanded an compact) and Input.Toggle which incles errorMessage and isRequired

## Sample Card
If appropriate, please include a link to a card in one of the samples directories that can be used to validate this change. This can be an existing card or a card added with this PR.
choice gif

## How Verified
How you verified the fix, including one or all of the following: 
1. New unit tests that were added if any. If none were added please add a quick line explaining why not.
2. Existing relevant unit/regression tests that you ran
3. Manual scenario verification if any; ***Do include .gif's or screenshots of the testing you performed here if you think that it 
will aid in code reviews or corresponding fixes on other platforms for eg.***

Unit tests added

### CODE REVIEW

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/TextInputRenderer.swift [hunk 1]
```diff
@@ -142,6 +142,12 @@ class TextInputRenderer: NSObject, BaseCardElementRendererProtocol {
     }
 }
 class ACRTextInputView: ACRTextField, InputHandlingViewProtocol {
+    var isRequired = false
+    
+    func showError() {
+        // hi
+    }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Renderers/TextInputRenderer.swift:149]: Hi !

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Views/ACRViewTests.swift [hunk 2]
```diff
@@ -174,8 +174,8 @@ class ACRViewTests: XCTestCase {
         
         view.addInputHandler(testinputHandler)
         view.handleSubmitAction(actionView: NSButton(), dataJson: nil, associatedInputs: true)
-        
-        XCTAssertEqual(0, actionDelegate.dict.count)
+        // Logic changed, so, isValid chack ignored unless 1.3 support enabled
+        XCTAssertEqual(1, actionDelegate.dict.count)
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Views/ACRViewTests.swift:178]: 👍

File: source/macos/AdaptiveCards/AdaptiveCardsTests/Views/ACRViewTests.swift [hunk 3]
```diff
@@ -357,6 +357,11 @@ private class FakeInputHandlingView: NSView, InputHandlingViewProtocol {
     var key: String = NSUUID().uuidString
     var isValid: Bool = true
     var errorMessageHandler: ErrorMessageHandlerDelegate?
+    var isRequired = false
+    var expectation: XCTestExpectation?
+    func showError() {
+        expectation?.fulfill()
+    }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCardsTests/Views/ACRViewTests.swift:364]: Probably a `showErrorCalled: Bool` flag will do here?
Or is there any async operation involved here?

Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCardsTests/Views/ACRViewTests.swift:364]: Yeah true, I just saw this line a few tests above it and I wanted to use it as it looked cool😂

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceSetView.swift [hunk 4]
```diff
@@ -81,15 +87,7 @@ class ACRChoiceSetView: NSView, InputHandlingViewProtocol {
...
-    }
-    
-    var isRequired: Bool {
-        return false
-    }
-    
-    func showError() {
-        // do nothing
+        return isRequired ? stackview.arrangedSubviews.contains(where: { let elem = $0 as? ACRChoiceButton; return elem?.state == .on }) : true
     }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceSetView.swift:91]: Not sure, do you need `isBasicValidationSatidfied` here?

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceSetView.swift [hunk 5]
```diff
@@ -140,14 +154,6 @@ class ACRChoiceSetCompactView: NSPopUpButton, InputHandlingViewProtocol {
...
-    }
-    
-    var isRequired: Bool {
-        return false
-    }
-    
-    func showError() {
-        // do nothing
+        return isRequired ? (arrayValues[indexOfSelectedItem] != nil) : true
     }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceSetView.swift:158]: Same here

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift [hunk 6]
```diff
@@ -226,18 +231,58 @@ class ACRContentStackView: NSView, ACRContentHoldingViewProtocol, SelectActionHa
...
+        addArrangedSubview(view)
+        if view is ACRContentStackView {
+            view.widthAnchor.constraint(equalTo: widthAnchor).isActive = true
+        }
+        setupErrorMessage(element: element, view: view)
+    }
+    
+    private func setupLabel(for element: ACSBaseInputElement) {
+        guard renderConfig.supportsSchemeV1_3, inputLabelField?.stringValue != element.getLabel() else { return }
+        if let label = element.getLabel(), !label.isEmpty {
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift:255]: `let label = element.getLabel(), !label.isEmpty` these can be clubed in the guard as well?

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift [hunk 7]
```diff
@@ -226,18 +231,58 @@ class ACRContentStackView: NSView, ACRContentHoldingViewProtocol, SelectActionHa
...
+            let labelView = staticTextField()
+            labelView.attributedStringValue = attributedString
+            addArrangedSubview(labelView)
+            setCustomSpacing(spacing: 3, after: labelView)
+            inputLabelField = labelView
+        }
+    }
+    
+    func setupErrorMessage(element: ACSBaseInputElement, view: NSView) {
+        guard renderConfig.supportsSchemeV1_3, errorMessageField?.stringValue != element.getErrorMessage() else { return }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRContentStackView.swift:274]:
```suggestion
    private func setupErrorMessage(element: ACSBaseInputElement, view: NSView) {
        guard renderConfig.supportsSchemeV1_3, errorMessageField?.stringValue != element.getErrorMessage() else { return }
```
Maybe can be private now

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/BaseCardElementRenderer.swift [hunk 8]
```diff
@@ -36,26 +36,13 @@ class BaseCardElementRenderer {
...
-            let attributedString = NSMutableAttributedString(string: label)
-            if let colorHex = hostConfig.getForegroundColor(style, color: .default, isSubtle: false), let textColor = ColorUtils.color(from: colorHex) {
-                attributedString.addAttributes([.foregroundColor: textColor], range: NSRange(location: 0, length: attributedString.length))
+        if let inputElement = element as? ACSBaseInputElement {
+            updatedView.configureInputElements(element: inputElement, view: view)
+        } else {
+            updatedView.addArrangedSubview(view)
+            if view is ACRContentStackView {
+                view.widthAnchor.constraint(equalTo: updatedView.widthAnchor).isActive = true
             }
```
Comment (blue-stone99) [inline source/macos/AdaptiveCards/AdaptiveCards/Renderers/BaseCardElementRenderer.swift:45]: Should this be in else?
Seems something to be common for all irrespective of ACSBaseInputElement

**Review by blue-stone99 — APPROVED** [review]
Minor comments. LGTM otherwise!

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceSetView.swift (hunk 4 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceSetView.swift:91]: No Akaash, in choiceset like we discussed, people can make the value empty string and for input.toggle it either submits true/falssse, never empty

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/BaseCardElementRenderer.swift (hunk 8 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCards/Renderers/BaseCardElementRenderer.swift:45]: For ACSBaseInputElement I add it in ACRContentStackView as for that we need to add error message below i, else we will have to again call another function from here to setup errorMessage

**Pull request merged**